Add new devicePixel ref test

greggman opened the pull request:
What the test does:

It sets the body to have a 2x2 checkerboard pattern fit to devicePixels (1 pixel in the pattern = 1 devicePixel)

It then puts a bunch of children above that in rows and uses `devicePixelContentBoxSize` to compute where each child is. It sets the same pattern for each child and offsets the pattern based on the child's calculated position so the patterns should align. If they don't align because the values returned from `devicePixelContentBoxSize` it should be very clear.

Here's Chrome running the test

<img width="692" alt="スクリーンショット 2022-07-01 午後6 28 34" src="https://user-images.githubusercontent.com/234804/176981813-67838b4a-2192-412e-b49a-0ea7a1486b92.png">


Here's Firefox running the test

<img width="691" alt="スクリーンショット 2022-07-01 午後6 27 50" src="https://user-images.githubusercontent.com/234804/176981810-6239c66c-f531-4fe0-8575-44088f464604.png">



Related:
https://github.com/web-platform-tests/wpt/pull/34569

greggman committed:
Add new devicePixel ref test

See comments in test for details

atotic reviewed (commented):
I will defer to dholbert on this one. I no longer work on web standards, and have used up my "free review time" for a while.

chrishtr commented on resize-observer/devicepixel2.html:
```diff
@@ -0,0 +1,157 @@
+
+const observer = new ResizeObserver(setPatternInEntries);
+observer.observe(backElem, {box:"device-pixel-content-box"});
+for (let i = 2; i < 15; ++i) {
```
Can you just achieve the same goal with a minimal number of elements declared in HTML rather than built in script? That will be a better reduced testcase, and also make it more clear what layout you're achieving

dholbert replied:
> The flexbox tests this indirectly (because varying number of children varies their locations/size).
[...]
>  I'd prefer to explicitly set location/size.

So, I don't think it's possible to both use this test's same relatively-clean "checkerboard overlaying" strategy (which I like) and to **also** directly/explicitly set the position of any element.  If you directly set an element's position to anything other than the origin, then you can't know for sure whether that element will be painted at an even vs. odd device-pixel-position.  And you have to have that information in order to properly set `background-position` on that element (or its children, etc) to align their checkerboard-image-background with the document's checkerboard-image-background.

(That's kind of the point of this test -- if the cumulative `devicePixelBoxSize` measurements of assumed-to-be-adjacent elements are correct, then we'll be able to correctly predict how to set `background-position` on the various checkerboard elements so that they align properly with the checkerboard behind them and won't stand out).

IMO this setup is pretty reasonable.  I could imagine (but probably wouldn't require) simplifying things by making all the flex containers have only 2 flex-items, both inflexible (`flex: none` with specified `width` and `height`), with a variety of widths on the first items in the flex containers (e.g. ranging from 10.0px, 10.1px, etc, up to 11px), so that the second flex item will end up snapping in different directions between different flex containers.  Then we would set the `background-position` on the second flex item based on its assumed device-pixel x/y coordinates, which we would determine based on the previous flex item's observed size and based on the cumulative sizes of the previous flex containers (as the test currently does).

This would let us use explicit hardcoded sizes, and it would sort-of let us explicitly specify positions as well (via positions being trivially inferrable based on the sizes of prior-siblings).

(We could also achieve a similar setup using floats instead of flexbox, but IMO float layout is hairier than flexbox in simple setups like this, particularly when we're using flexbox pretty trivially as this test does with no complex content-based sizing or alignment.)

dholbert reviewed (changes requested):
sorry for the delay here!

Overall, this makes sense to me, as long as we have the guarantee that two adjacent flex items should be drawn with directly-abutting device-pixel edges on their content boxes (which seems like a reasonable assumption).

I've got some readability/clarity notes at this point, but the overall approach seems reasonable to me.

dholbert commented on resize-observer/devicepixel2-ref.html:
```diff
@@ -0,0 +1,29 @@
+}
+setBackgroundPatternTo1DevicePixel();
+window.addEventListener('resize', setBackgroundPatternTo1DevicePixel);
+</script>
```
to clarify, this is the "angry red icon" I'm talking about -- it only shows up in some github UI, e.g. on the direct view of the commit itself:
![image](https://user-images.githubusercontent.com/426803/179086989-77dd54b2-e9b4-420e-83b7-bb6198efdfe9.png)

greggman replied:
fixed

dholbert commented on resize-observer/devicepixel2.html:
```diff
@@ -0,0 +1,157 @@
+
+const elemToEntry = new Map();
+
+function setPatternInEntries(entries) {
```
"setPatternInEntries" is a bit of a confusing name, since we're not putting the pattern in the entries.

Maybe rename this to "setPatternsUsingSizeInfo" or something along those lines?  That seems to be what this does, rather than doing something "in the entries".

greggman replied:
done

dholbert commented on resize-observer/devicepixel2.html:
```diff
@@ -0,0 +1,157 @@
+    });
+  }
+
+  const setPatternInEntry = function(elem, devicePixelX, devicePixelY) {
+    const entry = elemToEntry.get(elem);
```
(er, I guess there are two callsites; but the second callsite doesn't actually need this lambda and could be simplified to just `setPattern(backElem,0,0)` given that we don't actually need the width or height.  And as noted elsewhere, `backElem` might want to just disappear entirely for improved consistency with the reference case.)

greggman replied:
fixed

dholbert commented on resize-observer/devicepixel2.html:
```diff
@@ -0,0 +1,157 @@
+    top: 0;
+  }
+  .outer {
+    width: 99.5px;
```
99.5px feels unnecessarily-magic; consider changing this to 100px (or even 127px [same ballpark, but prime].)

I think your intent is that 99.5px is "weird" and likely to produce bugginess, but 100px actually produces more fully-black areas (and it does so sooner, i.e. in higher-up rows, than 99.5px does on my system at least -- Firefox Nightly with devicePixelRatio = 1)

The fractional size here implicitly suggests that we need a fractional size to trigger the bug in question; but in fact we don't.

greggman replied:
fixed

dholbert commented on resize-observer/devicepixel2.html:
```diff
@@ -0,0 +1,157 @@
+    display: flex;
+  }
+  .outer>* {
+    flex: 1 1 auto;
```
Let's just condense this to `flex: 1;` -- that's essentially equivalent (and simpler). The trailing "1 auto" is irrelevant and unused in practice, since the matched "inner" elements don't have any contents and so their "auto" size is 0 anyway.

(See https://drafts.csswg.org/css-flexbox-1/#flex-common which has the spec's suggested abbreviated usages for the `flex` shorthand, with "flex: <number>" being the preferred way to divide-the-free-space-equally.)

greggman replied:
done

greggman commented on resize-observer/devicepixel2-ref.html:
```diff
@@ -0,0 +1,29 @@
+
+document.body.style.backgroundImage = `url("${create2x2CheckerboardDataURL}")`;
+
+const px = v => `${v}px`;
```
fixed

greggman commented on resize-observer/devicepixel2-ref.html:
```diff
@@ -0,0 +1,29 @@
+<!doctype html>
+<body>
+</body>
+<script>
```
Fixed. But just FYI, I created this file by copying [`devicepixel.html`](https://github.com/greggman/wpt/blob/master/resize-observer/devicepixel.html)

greggman commented on resize-observer/devicepixel2.html:
```diff
@@ -0,0 +1,157 @@
+    devicePixelY += outerHeight;
+  }
+
+  setPatternInEntry(backElem, 0, 0);
```
done

greggman commented on resize-observer/devicepixel2.html:
```diff
@@ -0,0 +1,157 @@
+const observer = new ResizeObserver(setPatternInEntries);
+observer.observe(backElem, {box:"device-pixel-content-box"});
+for (let i = 2; i < 15; ++i) {
+  const elemToSizes = new Map();
```
fixed

greggman commented on resize-observer/devicepixel2.html:
```diff
@@ -0,0 +1,157 @@
+const frontElem = document.querySelector('#front');
+const backElem = document.querySelector('#back');
+
+const elemToEntry = new Map();
```
fixed.

greggman commented on resize-observer/devicepixel2.html:
```diff
@@ -0,0 +1,157 @@
+}();
+const patternURL = create2x2CheckerboardDataURL();
+
+function setPattern(elem, x, y, width, height) {
```
done

greggman commented on resize-observer/devicepixel2-ref.html:
```diff
@@ -0,0 +1,29 @@
+  document.body.style.backgroundSize = `${twoDevicePixels}px ${twoDevicePixels}px`;
+}
+setBackgroundPatternTo1DevicePixel();
+window.addEventListener('resize', setBackgroundPatternTo1DevicePixel);
```
done

greggman commented:
Thank you for the review. I think I addressed all the issues you raised. 

I changed the pattern to be 4x4 instead of 2x2. My thinking was 2x2 will fail if off by 1 but succeed if off by 2 or if both x and y are off by 1.  With 4x4 the pattern should fail in more cases (though I expect the normal "bad" case to be off by 1)

dholbert commented on resize-observer/devicepixel2.html:
```diff
@@ -0,0 +1,163 @@
+  }
+  #front {
+    position: absolute;
+    left: 0;
```
Is there a reason we're bothering with `front` here at this point?  Can we just get rid of it and append to `body` instead, to simplify a bit more?

greggman replied:
so I just tried this and the bug does not repo in firefox if I append to body. I don't know if that still means the test should append to body or not. I honestly don't know the intent of the WPT. In WebGL conformance tests (which I'm very familiar with), we check in tests all the time that provoke bugs in individual browsers.

dholbert replied:
> so I just tried this and the bug does not repo in firefox if I append to body

Thanks for testing that! Definitely want to be sure the bug still repro's in affected versions.

Can you just remove these #front styles? They should be unnecessary, and I confirmed I can still repro the bug in Firefox release with that change.  (Doesn't seem to make a difference.)

(I started to test appending to the body, but that's complicated by the fact that the we iterate over the elements using `children`, and the `<script>` element is one of the body's children right now. :))

dholbert commented on resize-observer/devicepixel2.html:
```diff
@@ -0,0 +1,163 @@
+  }
+  frontElem.appendChild(outer);
+}
+</script>
```
nit: add newline at end of file

(I think you fixed this for the reference case but not for the testcase)

greggman replied:
done

dholbert commented on resize-observer/devicepixel2.html:
```diff
@@ -0,0 +1,163 @@
+  const r = [255, 0, 0, 255];
+  const g = [0, 255, 0, 255];
+
+  const imageData = new ImageData(4, 4);
```
s/4/patternSize/ here maybe, since we've got that named constant?

greggman replied:
done

dholbert reviewed (commented):
Thanks for addressing the review feedback!  I'll take a further look later and run this against Firefox with the patch (the fix for our bug landed today!)

Dropping some quick nits for now...

dholbert commented:
Viewing today's updated test-version in Chrome on my regular no-pixel-scaling display, I see "fringes" around various flex items at a handful of zoom levels -- e.g. 110% and 125% at least (the first two Ctrl+ ticks).

Needless to say, those location-specific "fringes" aren't present in the reference case, because the reference case just draws a single pattern across the whole page.  So, the test seems to fail in Chrome under full-page zoom (and presumably at various HiDPI display levels).

Assuming a valid test and implementation, I think we wouldn't expect these fringes to be present... Full-page-zoom does get reflected in `devicePixelRatio`, so the test should theoretically be accounting for it properly in its device pixel arithmetic.  I wonder if this is a test bug (/ if we're assuming too much) or a Chrome bug.

greggman commented:
wow, that was hard to see. It was a bug in the test. The pattern needed to be offset `-(x % patternSize)`. That bug didn't show when the pattern was 2x2 since positive or negative would be the same but with a 4x4 pattern the offset needs to be negative. Uploaded a fix and I changed the pattern so the issue would have stuck out more.  Also note: Firefox fixed their bug yesterday so Firefox Nightly display this correctly now.

dholbert commented on resize-observer/devicepixel2.html:
```diff
@@ -0,0 +1,163 @@
+const patternURL = createPatternDataURL();
+
+/**
+ * Set the pattern's size on this element so that it draws were
```
s/were/where/

dholbert commented on resize-observer/devicepixel2.html:
```diff
@@ -0,0 +1,163 @@
+ * its position so it's aligned to the pattern of the body element.
+ * The pattern is 'patternSize' large. x and y are the computed 
+ * devicePixel position of the element. To make the pattern align
+ * we need to set its position to -(x % patternSize), -(y % patternSize)
+ * in devicePixels
+*/
+function setPattern(elem, x, y) {
+  const oneDevicePixel = 1 / devicePixelRatio;
+  const patternPixels = oneDevicePixel * patternSize;
+  const xOff = -(x % patternSize);
+  const yOff = -(y % patternSize);
```
(optional): I think you could set the background-position (xOff and yOff) to just `-x -y` in device pixels (no modulus operation needed), right?

Essentially, we'd be setting every element's background-position to 0 0 in the "global coordinate system (which for an element at `x,y` is `-x, -y`).  Then it will be trivially aligned/properly-overlaid, as long as we have the correct device-pixel positioning (i.e. as long as we don't have the Firefox bug).

The current modulus setup works too, but feels slightly more complex / less intuitive, IMO.

greggman replied:
done

dholbert commented on resize-observer/devicepixel2.html:
```diff
@@ -0,0 +1,163 @@
+thirdChildLeft = entry(firstChild).devicePixelContentBoxSize[0].inlineSize + 
+                 entry(secondChild).devicePixelContentBoxSize[0].inlineSize
+
+If the element's left edge mod 4 is an non-zero value then we need to set the
+background-position of the element's pattern left edge mod 4 devicePixels.
+
+If what devicePixelContentBox reports is correct the children's patterns
+will align with the pattern in the body and the children will be invisible.
+If what devicePixelContentBox reports is incorrect the patterns will not
+align and it should be clearly visible.
```
This comment needs a rewrite, if you take my above suggestion.

greggman replied:
done

dholbert commented on resize-observer/devicepixel2.html:
```diff
@@ -0,0 +1,163 @@
+  const yOff = -(y % patternSize);
+  elem.style.backgroundImage = `url("${patternURL}")`;
+  elem.style.backgroundSize = `${patternPixels}px ${patternPixels}px`;
+  elem.style.backgroundPosition = `${xOff * oneDevicePixel}px ${yOff * oneDevicePixel}px`;
```
(optional): I think you could set the background-position to `-x -y` in device pixels (i.e. `${xOff * oneDevicePixel}px)` and skip the modulus operations, and get rid of xOff/yOff?

Essentially, we can just set every element's background-position to 0 0 in the "global coordinate system (and for an element at `x,y`, the relative position of the global origin is `-x, -y`).

Then every element's background-pattern will be trivially aligned with the body's background, as long as we have the correct x/y positions here (which is the bug we're testing).

The current modulus setup works too, but feels slightly more complex / less intuitive, IMO.

(I confirmed that I can still repro the bug in Firefox release with `s/xOff/x/` and `s/yOff/y/` here, too, BTW.)

greggman replied:
done

dholbert commented on resize-observer/devicepixel2-ref.html:
```diff
@@ -0,0 +1,41 @@
+
+  const imageData = new ImageData(4, 4);
+  imageData.data.set([
+    b, t, r, t,
+    t, b, t, g,
+    r, t, b, t,
+    t, g, t, b,
```
You've got a different pixel-pattern here vs. in the testcase, I think by accident? This makes the test fail in all browsers for un-interesting reasons right now. :)  (Though it's hard to see without e.g. using an accessibility tool to scale up the device pixels.)

If I fix this (e.g. using the testcase's setup in the reference case), then I confirmed that the test passes for me in the WPT harness (including with the other changes I suggested -- using `-x -y` in device pixels and removing the `#front` css rule)

greggman replied:
done

greggman commented:
Thanks again for the review.  I think I addressed your latest comments.

dholbert commented:
Thanks for the fixes!

One more thing to fix: there's a "lint" task that's failing, due to some whitespace-at-the-end-of-various-lines (see the `lint` task in the `Some checks were not successful` section here -- direct link: https://github.com/web-platform-tests/wpt/pull/34682/checks?check_run_id=7383242439

That needs fixing. With that, this might be good, I think? I'll take a final look tomorrow and let you know if I see any final issues. :)

FWIW, I also ran a (slightly tweaked) version of your test (as-of-earlier-today) through Firefox Nightly CI, and I confirmed that it passes on trunk -- e.g. here's the pass result for a MacOS run:
https://treeherder.mozilla.org/logviewer?job_id=384603288&repo=try&lineNumber=19899-19902
...and I'm doing another Try run to confirm that it fails on builds from before the patch landed (which I expect it does based on observing it rendering badly in Firefox release):
https://treeherder.mozilla.org/#/jobs?repo=try&revision=e178b200151fdfe17056da1e7c5cd99da8c53ed3

dholbert commented:
(There's also a failing task in "Azure Pipelines" but that seems to be some sort of infrastructure issue with Safari Technical Preview failing to be downloaded for a particular platform -- not your problem / not related to this test, I think.

Direct link to the failure there:
https://dev.azure.com/web-platform-tests/b14026b4-9423-4454-858f-bf76cf6d1faa/_apis/build/builds/84592/logs/27
```
2022-07-18T04:53:00.0501510Z + ./wpt install --channel preview --download-only -d . --rename STP safari browser
2022-07-18T04:53:07.8874680Z INFO:install:Now installing safari browser...
[...]
2022-07-18T04:53:08.1433370Z ValueError: no download for 12.4
2022-07-18T04:53:08.1621610Z ##[error]Bash exited with code '1'.
2022-07-18T04:53:08.1715030Z ##[section]Finishing: Install Safari Technology Preview
```

@jgraham do you know who we should notify about that? (Perhaps someone who works on WPT infra stuff and/or Safari/WPT integration?)

dholbert commented:
(maybe @gsnedders knows what's going on with the Safari Technology Preview azure-pipelines task, and if that's something we need to worry about? [seems like an infrastructure issue, as noted above]

gsnedders commented:
> (maybe @gsnedders knows what's going on with the Safari Technology Preview azure-pipelines task, and if that's something we need to worry about? [seems like an infrastructure issue, as noted above]

#34875

dholbert commented on resize-observer/devicepixel2-ref.html:
```diff
@@ -0,0 +1,41 @@
+  const r = [255, 0, 0, 255];
+  const g = [0, 255, 0, 255];
+
+  const imageData = new ImageData(4, 4);
```
`s/4/patternSize/`

(for consistency with the testcase, where I think you fixed this previously)

dholbert commented on resize-observer/devicepixel2.html:
```diff
@@ -0,0 +1,152 @@
+<!doctype html>
+<link rel="match" href="devicepixel2-ref.html">
+<meta name="assert" content="Device pixel content box sizes and pixel snapping are correct in Resize Observer callback">
```
As atotic noted, "correctness" is sort of undefined for pixel-snapping, so it feels a little like over-promising to say that we're checking for "correct" pixel-snapping.

Maybe better/more-precise to say something like this:
> "Resize Observer's reported device pixel content box size should be consistent with the actual pixel-snapped painted box size"

dholbert commented on resize-observer/devicepixel2.html:
```diff
@@ -0,0 +1,152 @@
+thirdChildLeft = entry(firstChild).devicePixelContentBoxSize[0].inlineSize +
+                 entry(secondChild).devicePixelContentBoxSize[0].inlineSize
+
+By setting the element's pattern's position -leftEdge that should align
+the pattern to 0 on the page.
```
This sentence doesn't quite make sense to me.  Maybe clearer:

> Using the children's device-pixel positions (determined as described above), we then set each child's background-position to the negated version of the position that we computed for that child.  This effectively makes its pattern's origin match the document origin.

dholbert commented on resize-observer/devicepixel2.html:
```diff
@@ -0,0 +1,152 @@
+</style>
+<body>
+  <div></div>
+</body>
+<script>
```
nit: let's move the closing </body> to the end (as you did for the reference case), to reflect the reality of the actual DOM structure here.

dholbert commented on resize-observer/devicepixel2.html:
```diff
@@ -0,0 +1,152 @@
+
+const frontElem = document.querySelector('div');
+
+const elemToDevicePixelSize = new Map();
+
+function setPatternsUsingSizeInfo(entries) {
```
maybe move the `elemToDevicePixelSize` declaration inside of the `setPatternsUsingSizeInfo()` function, since that's the only place it's used?

dholbert commented on resize-observer/devicepixel2.html:
```diff
@@ -0,0 +1,152 @@
+It has the first `<div>` starting at the top left corner of the page, the
+same as the body so they should both have a devicePixel position of 0,0
+
+It then sets a repeating background-image and background-size on the
+body element and all of the .outer children elements so they have a 4x4
+checkerboard image at a size that should draw each pixel of the 4x4
+checkerboard pattern in single devicePixels.
+
+The devicePixelContentBoxSize of all the elements is queried with
+a ResizeObserver.
+
+Finally the background-position of each child element is set so that
+the patterns should align.
```
This chunk of the comment doesn't match what the test actually does.

In particular, three things to address:
 - The phrase ".outer children" is ambiguous & sounds like it's referring to the .outer elements themselves (it sounds like "children who have class .outer"); and that's confusing because the .outer elements don't get any background at all.  Maybe just rephrase as "inner children"? (since you're talking about the innermost things)

 - The current wording suggests that we set the background-position in a final pass (with the image + size having been set earlier).  But in fact, we set the pattern/size/position all at once.

 - The current wording suggests that we set the body's background first and then handle all of the elements later.  But in fact, we handle the body *last*.

Could you update the wording here to address these?  (In the case of the last item, we could just move the `setPattern(document.body, 0, 0);` to be the first thing that happens in the ResizeObserver callback. It is useful to keep it in the callback, to ensure that that the test will fail if the callback never fires.)

dholbert reviewed (commented):
Final clarity nits; looks good to me with these changes!

dholbert commented on resize-observer/devicepixel2.html:
```diff
@@ -0,0 +1,152 @@
+  }
+
+  let devicePixelY = 0;
+  for (const outerElem of frontElem.children) {
```
Sorry, one more thing...  It occurs to me that "outer" and "front" are sort of confusing names at this point.  In particular, the seem ~backwards:
 - "outer" is not the outermost element, but rather "front" is the outermost thing.
 - "front" is the container for all of the page's content, which puts it in **back** of all the page's content (not really "in front")

Maybe clearer if we renamed "outer" to "line" or "row", and then rename "front" to "wrapper"?

dholbert reviewed (commented):
Looks good, aside from two small typos - thanks for all the fixups & getting this over the finish line! I think we can merge once these last nits are addressed.

dholbert commented:
I clicked the button to merge; hopefully everything works properly. :)

Thanks again for the Firefox bug report and for the thorough test!

dholbert commented:
wpt.fyi now has this test deployed, and shows it passing on all browsers; hooray!
https://wpt.fyi/results/resize-observer/devicepixel2.html
